test/integration/listener_lds_integration_test.cc

```diff
@@ -1356,6 +1356,146 @@ TEST_P(ListenerFilterIntegrationTest, UpdateListenerFilterOrder) {
+    // Add the static cluster to serve LDS.
+    auto* lds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+    lds_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+    lds_cluster->set_name("lds_cluster");
+    ConfigHelper::setHttp2(*lds_cluster);
+  });
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    listener_config_.Swap(bootstrap.mutable_static_resources()->mutable_listeners(0));
+    listener_config_.set_name("listener_foo");
+    listener_config_.mutable_enable_reuse_port()->set_value(false);
+    ENVOY_LOG_MISC(debug, "listener config: {}", listener_config_.DebugString());
+    bootstrap.mutable_static_resources()->mutable_listeners()->Clear();
+    auto* lds_config_source = bootstrap.mutable_dynamic_resources()->mutable_lds_config();
+    lds_config_source->set_resource_api_version(envoy::config::core::v3::ApiVersion::V3);
+    auto* lds_api_config_source = lds_config_source->mutable_api_config_source();
+    lds_api_config_source->set_api_type(envoy::config::core::v3::ApiConfigSource::GRPC);
+    lds_api_config_source->set_transport_api_version(envoy::config::core::v3::V3);
+    envoy::config::core::v3::GrpcService* grpc_service = lds_api_config_source->add_grpc_services();
+    setGrpcService(*grpc_service, "lds_cluster", fake_upstreams_[1]->localAddress());
+  });
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "1");
+  };
+  use_lds_ = false;
+  initialize();
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  test_server_->waitUntilListenersReady();
+  // NOTE: The line above doesn't tell you if listener is up and listening.
+  test_server_->waitForCounterGe("listener_manager.listener_create_success", 1);
+  // Workers not started, the LDS added test_listener is in active_listeners_ list.
+  EXPECT_EQ(test_server_->server().listenerManager().listeners().size(), 1);
+  registerTestServerPorts({"test_listener"});
+  std::string data = "hello";
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("test_listener"));
+  ASSERT_TRUE(tcp_client->write(data));
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  ASSERT_TRUE(fake_upstream_connection->waitForData(data.size(), &data));
+  tcp_client->close();
+
+  auto* socket_option = listener_config_.add_socket_options();
+  socket_option->set_level(IPPROTO_IP);
+  socket_option->set_name(IP_TOS);
+  socket_option->set_int_value(8); // IPTOS_THROUGHPUT
+  socket_option->set_state(envoy::config::core::v3::SocketOption::STATE_LISTENING);
+  ENVOY_LOG_MISC(debug, "listener config: {}", listener_config_.DebugString());
+  sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "2");
+  test_server_->waitForCounterGe("listener_manager.listener_create_failure", 1);
```

Comment:
Is there a way I can't ensure the listener update reject for a specific reason?

Reply:
Perhaps a better way would be to use https://github.com/envoyproxy/envoy/blob/main/source/server/listener_hooks.h and add a new hook for a failure reason that you can trap in tests? cc @alyssawilk to see if she has any other thoughts about the test part of this change.

Reply:
This results in the config being rejected yes?  If so that should always come with a warning / error log explaining why the config is invalid, and you can use EXPECT_LOG_CONTAINS to make sure you got the cause you expected.

Reply:
I can use `EXPECT_LOG_CONTAINS` to get expected message, like
```
EXPECT_LOG_CONTAINS(
      "warning",
      "gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error "
      "adding/updating listener(s) listener_foo: Listener listener_foo: doesn't support update any "
      "socket options when the reuse port isn't enabled",
      test_server_->waitForCounterGe("listener_manager.lds.update_rejected", 1))
```
but there isn't a point I can ensure the log already emit

I wait on the stat `listener_manager.lds.update_rejected`, but I can't ensure the log emitted. Even I change the code to emit the log first, I still see the test flaky.
https://github.com/envoyproxy/envoy/blob/ad186c448cf8c5e78454d2adf5f228e9c90a997c/source/common/config/grpc_subscription_impl.cc#L127-L128

Maybe I still need to use listener hook.

Reply:
Emm.. I can put something more into the EXPECT_LOG_CONTAINS

```
EXPECT_LOG_CONTAINS(
      "warning",
      "gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error "
      "adding/updating listener(s) listener_foo: Listener listener_foo: doesn't support update any "
      "socket options when the reuse port isn't enabled",
      {
        sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "2");
        test_server_->waitForCounterGe("listener_manager.lds.update_rejected", 1);
        IntegrationTcpClientPtr tcp_client2 = makeTcpConnection(lookupPort("test_listener"));
        ASSERT_TRUE(tcp_client2->write(data));
        FakeRawConnectionPtr fake_upstream_connection2;
        ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection2));
        ASSERT_TRUE(fake_upstream_connection2->waitForData(data.size(), &data));
        tcp_client2->close();
      }
  );
```

Then I got enough time to wait the log emitted, I think that is enough